cockatrice/src/handle_public_servers.cpp

```diff
@@ -0,0 +1,92 @@
+//
+// Created by Zachary Halpern on 4/6/18.
+//
+
+#include "handle_public_servers.h"
+#include "qt-json/json.h"
+#include "settingscache.h"
+#include <QUrl>
+
+#define PUBLIC_SERVERS_JSON "https://cockatrice.github.io/public-servers.json"
+
+HandlePublicServers::HandlePublicServers(QObject *parent) : QObject(parent), nam(nullptr), reply(nullptr)
+{
+}
+
+void HandlePublicServers::downloadPublicServers()
+{
+    QUrl url(QString(PUBLIC_SERVERS_JSON));
+    nam = new QNetworkAccessManager(this);
+    reply = nam->get(QNetworkRequest(url));
+    connect(reply, SIGNAL(finished()), this, SLOT(actFinishParsingDownloadedData()));
+}
+
+void HandlePublicServers::actFinishParsingDownloadedData()
+{
+    reply = dynamic_cast<QNetworkReply *>(sender());
+    QNetworkReply::NetworkError errorCode = reply->error();
+
+    // After finished() is called, this object will be deleted
+    nam->deleteLater();
+
+    if (errorCode == QNetworkReply::NoError) {
+        // Get current saved hosts
+        UserConnection_Information uci;
+        savedHostList = uci.getServerInfo();
+
+        // List of public servers
+        // Will be populated below
+        QStringList publicServersAvailable;
+        QStringList publicServersToRemove;
+
+        // Downloaded data from GitHub
+        QString jsonData = QString(reply->readAll());
+        auto jsonMap = QtJson::Json::parse(jsonData).toMap();
+
+        // Servers available
+        auto publicServersJSONList = jsonMap["servers"].toList();
```

Comment:
Can you split this behavior out into its own function taking jsonMap as a parameter? That would make this class far more testable because we can ignore all the network logic and focus on testing the json logic.

Reply:
Done